pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java

```diff
@@ -949,6 +949,48 @@ The max allowed delay for delayed delivery (in milliseconds). If the broker rece
             + " back and unack count reaches to `limit/2`. Using a value of 0, is disabling unackedMessage-limit"
             + " check and broker doesn't block dispatchers")
     private int maxUnackedMessagesPerBroker = 0;
+
+    @FieldContext(
+            category = CATEGORY_POLICIES,
+            doc = "For Key_Shared subscriptions, if messages cannot be dispatched to consumers due to a slow consumer"
+                    + " or a blocked key hash (because of ordering constraints), the broker will continue reading more"
+                    + " messages from the backlog and attempt to dispatch them to consumers until the number of replay"
+                    + " messages reaches the calculated threshold.\n"
+                    + "Formula: threshold = max(keySharedLookAheadMsgInReplayThresholdPerConsumer *"
+                    + " connected consumer count, keySharedLookAheadMsgInReplayThresholdPerSubscription)"
+                    + ".\n"
+                    + "Setting this value to 0 will disable the limit calculated per consumer.",
+            dynamic = true
+    )
+    private int keySharedLookAheadMsgInReplayThresholdPerConsumer = 1000;
+
+    @FieldContext(
+            category = CATEGORY_POLICIES,
+            doc = "For Key_Shared subscriptions, if messages cannot be dispatched to consumers due to a slow consumer"
+                    + " or a blocked key hash (because of ordering constraints), the broker will continue reading more"
+                    + " messages from the backlog and attempt to dispatch them to consumers until the number of replay"
+                    + " messages reaches the calculated threshold.\n"
+                    + "Formula: threshold = max(keySharedLookAheadMsgInReplayThresholdPerConsumer *"
+                    + " connected consumer count, keySharedLookAheadMsgInReplayThresholdPerSubscription)"
+                    + ".\n"
+                    + "This value should be set to a value less than 2 * managedLedgerMaxUnackedRangesToPersist.\n"
+                    + "Setting this value to 0 will disable the limit calculated per subscription.\n",
+            dynamic = true
+    )
+    private int keySharedLookAheadMsgInReplayThresholdPerSubscription = 10000;
+
+
+    @FieldContext(
+            category = CATEGORY_POLICIES,
+            doc = "For Key_Shared subscriptions, if messages cannot be dispatched to consumers due to a slow consumer"
+                    + " or a blocked key hash (because of ordering constraints), the broker will continue reading more"
+                    + " messages from the backlog and attempt to dispatch them to consumers until the number of replay"
+                    + " messages reaches the calculated threshold.\n"
+                    + "This setting controls whether look ahead is enabled when recently joined consumers are present.",
+            dynamic = true
+    )
+    private boolean keySharedLookAheadEnabledWhenRecentlyJoinedConsumersExist = false;
```

Comment:
Is this going to change existing behavior, with the default setting?

Reply:
no, that's what exists in the current released versions that it doesn't "look ahead" when recently joined consumers exist. I guess we need to experiment whether this could be changed to do the look ahead also in that case. I was cautious about changing the existing behavior and that's why I added this flag to control the behavior. We'd need to experiment and analyze whether the default value could be changed without risks of regressions. I'm trying to minimize the risks of regressions.

Reply:
Two things
- Even if `keySharedLookAheadEnabledWhenRecentlyJoinedConsumersExist` is `false`, the broker will still look ahead and read entries, because it only affects when `messagesSent == 0`, see https://github.com/apache/pulsar/pull/23231/files#diff-c48d5c94842ac8c9a0c9314b207298069f38c8dcfeda4a9886fb3bb1f77843f2R362. 
- Since you set `keySharedLookAheadEnabledWhenRecentlyJoinedConsumersExist` to `false`, all tests will run with `keySharedLookAheadEnabledWhenRecentlyJoinedConsumersExist = true`, so we have no tests cover the cases that `keySharedLookAheadEnabledWhenRecentlyJoinedConsumersExist = false`.

Reply:
> Even if keySharedLookAheadEnabledWhenRecentlyJoinedConsumersExist is false, the broker will still look ahead and read entries, because it only affects when messagesSent == 0

Yes, I'll cover this in the documentation. That's the reason why I'm calling it "look ahead". I'll find a way to improve this.

Reply:
> * Since you set the default value of `keySharedLookAheadEnabledWhenRecentlyJoinedConsumersExist` to `false`, all tests will run with `keySharedLookAheadEnabledWhenRecentlyJoinedConsumersExist = false`, so we have no tests cover the cases that `keySharedLookAheadEnabledWhenRecentlyJoinedConsumersExist = true`.

Valid point. I'm planning to remove this and make look ahead enabled when recently joined consumers exist if the reason why a message couldn't be dispatched was caused by a blocked hash. This will make the solution more robust too.